[mlir][vector] Add support for linearizing Extract, ExtractStridedSlice, Shuffle VectorOps in VectorLinearize

charithaintc opened the pull request:
This PR adds support for converting `vector.extract_strided_slice` and `vector.extract` operations to equivalent `vector.shuffle` operations that operates on linearized (1-D) vectors.  `vector.shuffle` operations operating on n-D (n > 1) are also converted to equivalent shuffle operations working on linearized vectors.

charithaintc committed:
add linearize patterns for Extract, ExtractStridedSlice, Shuffle VectorOps

charithaintc committed:
Merge branch 'main' of https://github.com/llvm/llvm-project

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

charithaintc committed:
remove comments

charithaintc committed:
fix test

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
+    if (shuffleOp.getV1VectorType().isScalable() ||
+        shuffleOp.getV2VectorType().isScalable() ||
+        dstType.cast<VectorType>().isScalable())
+      return rewriter.notifyMatchFailure(loc,
+                                         "scalable vectors are not supported.");
```
Could you add a test to capture this? You should be able to re-use linearlize.mlir - just add `-verify-diagnostics` to the RUN line. Same for the other pattern.

charithaintc replied:
I tried this approach. But it does not result in any error reported to standard output. I can see the error if I use `debug-only=dialect-conversion`.

charithaintc replied:
Good point! #88733 is merged now. so replaced with assert.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
+  unsigned targetVectorBitWidth;
+};
+
+struct LinearizeVectorExtractStridedSlice final
```
Please add comments summarising this transformation. Would be good to have that for other patterns too.

charithaintc replied:
Added some comments describing what the rewrite pattern does.

charithaintc replied:
Thanks for pointing this out. I tried my best to add some example based comments for each pattern.

banach-space commented on mlir/test/Dialect/Vector/linearize.mlir:
```diff
@@ -153,3 +153,95 @@ func.func @test_0d_vector() -> vector<f32> {
+func.func @test_extract_strided_slice_2(%arg0 : vector<2x8x2xf32>) -> vector<1x4x2xf32> {
+  // DEFAULT: %[[ARG:.*]] = vector.shape_cast %[[ORIG_ARG]] : vector<2x8x2xf32> to vector<32xf32>
+  // DEFAULT: %[[SHUFFLE:.*]] = vector.shuffle %[[ARG]], %[[ARG]]
+  // DEFAULT: [20, 21, 22, 23, 24, 25, 26, 27] : vector<32xf32>, vector<32xf32>
```
What are these magic numbers?

charithaintc replied:
Hi @banach-space, these are the shuffle indices of the generated vector.shuffle. In this case we extract a `1x4x2` slice from the `2x8x2` vector. `[20, 21, 22, 23, 24, 25, 26, 27] ` represents the indices of the linearized source vector that needs to be shuffled to construct the extracted output vector. 

Should I include some comments with an example?

Hardcode84 replied:
I think check can be changed to `// DEFAULT-SAME:` to verify it belongs the same line with shuffle.

charithaintc replied:
added `// *-SAME` to all lines with shuffle indices.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +105,234 @@ struct LinearizeVectorizable final
+    }
+
+    auto mask = shuffleOp.getMask();
+    auto totalSize = mask.size() * shuffleSliceLen;
```
`totalSize` of what?

charithaintc replied:
fixed.

banach-space reviewed (commented):
Thanks for the contribution!

I've only been able to scan this very quickly - will take another look later, but mostly looks good. In the meantime, please make sure that you follow the guidelines. In particular:
* https://mlir.llvm.org/getting_started/Contributing/#commit-messages

charithaintc committed:
address comments

charithaintc commented:
> Thanks for the contribution!
> 
> I've only been able to scan this very quickly - will take another look later, but mostly looks good. In the meantime, please make sure that you follow the guidelines. In particular:
> 
> * https://mlir.llvm.org/getting_started/Contributing/#commit-messages

Hi @banach-space,

Thank you very much for the review. I have addressed the comments. Please let me know if there are any additional concerns.

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -139,9 +373,19 @@ void mlir::vector::populateVectorLinearizeTypeConversionsAndLegality(
 
-  patterns.add<LinearizeConstant, LinearizeVectorizable>(
+  patterns.add<LinearizeConstant, LinearizeVectorizable, LinearizeVectorShffle,
+               LinearizeVectorExtract, LinearizeVectorExtractStridedSlice>(
```
Could we move this to a different `populate` method? Flattening an elementwise op is potentially a no-op but generating shuffles for extract ops could have a noticeable impact in performance

charithaintc replied:
I added a new populate method called `populateVectorLinearizeToShuffleRewritePatterns`. If you have a better name suggestion for this please let me know.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
+
+/// This pattern converts the vector.shuffle operation that works on nD (n > 1)
+/// vectors to a vector.shuffle operation that works on linearized vectors.
+struct LinearizeVectorShffle final
```
Shffle -> Shuffle?

charithaintc replied:
fixed.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
+      unsigned targetVectBitWidth = std::numeric_limits<unsigned>::max(),
+      PatternBenefit benefit = 1)
+      : OpConversionPattern(typeConverter, context, benefit),
+        targetVectorBitWidth(targetVectBitWidth) {}
```
Thanks for adding support for the target vector bitwidth!

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
+    // here, extraction granularity is 8.
+    int64_t extractSliceLen = 1;
+    auto n = extractOp.getSourceVectorType().getRank();
+    auto k = (int64_t)offsets.size();
```
nit: we should spell out `auto` when the type is not literally redundant on the RHS.

charithaintc replied:
fixed.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -106,6 +108,238 @@ struct LinearizeVectorizable final
+        extractSliceLen *= extractOp.getSourceVectorType().getShape()[i + k];
+      }
+    }
+    // get total number of extracted slices
```
nit: we should use capital letters and periods in comments per coding guidelines.

charithaintc replied:
fixed all comments.

dcaballe reviewed (commented):
I did a first pass. LG! A few comments

charithaintc committed:
address comments

charithaintc commented:
> I did a first pass. LG! A few comments

Hi @dcaballe,

Thanks for the review. I have addressed all the comments.

charithaintc committed:
fix tests

Hardcode84 commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -145,3 +378,24 @@ void mlir::vector::populateVectorLinearizeTypeConversionsAndLegality(
+void mlir::vector::populateVectorLinearizeToShuffleRewritePatterns(
+    TypeConverter &typeConverter, RewritePatternSet &patterns,
+    ConversionTarget &target, unsigned int targetBitWidth) {
+  target.markUnknownOpDynamicallyLegal(
```
I think you can just do `target.addDynamicallyLegalOp<vector::ShuffleOp>(...)`

charithaintc replied:
fixed.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+  LogicalResult
+  matchAndRewrite(vector::ExtractStridedSliceOp extractOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto dstType = getTypeConverter()->convertType(extractOp.getType());
```
Why is this needed?

charithaintc replied:
yes. this is not needed. I removed the `dstType` checks.

dcaballe replied:
nit: spell out `auto`

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+        dstType.cast<VectorType>().isScalable())
+      return rewriter.notifyMatchFailure(loc,
+                                         "scalable vectors are not supported.");
+    if (!isLessThanTargetBitWidth(extractOp, targetVectorBitWidth))
+      return rewriter.notifyMatchFailure(
+          extractOp, "Can't flatten since targetBitWidth <= OpSize");
```
This would correspond to `BW-0` in your tests, right?

charithaintc replied:
yes.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+      return rewriter.notifyMatchFailure(
+          extractOp, "Can't flatten since targetBitWidth <= OpSize");
+
+    auto offsets = extractOp.getOffsets().getValue();
+    auto sizes = extractOp.getSizes().getValue();
+    auto strides = extractOp.getStrides().getValue();
```
```suggestion
    auto offsets = extractOp.getOffsets();
    auto sizes = extractOp.getSizes();
    auto strides = extractOp.getStrides();
```

charithaintc replied:
fixed.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+      return rewriter.notifyMatchFailure(
+          extractOp, "Strided slice with stride != 1 is not supported.");
+    Value srcVector = adaptor.getVector();
+    // If kD offsets are specified for nd source vector (n > k), the granularity
+    // of the extraction is greater than 1. In this case last (n-k) dimensions
+    // form the extraction granularity.
+    // example :
+    //  %0 = vector.extract_strided_slice %src { offsets = [0, 0], sizes = [2,
+    //  2],
+    //     strides = [1, 1]} : vector<4x8x8xf32> to vector<2x2x8xf32>
+    // here, extraction granularity is 8.
```
```suggestion
    // If kD offsets are specified for nD source vector (n > k), the granularity
    // of the extraction is greater than 1. In this case last (n-k) dimensions
    // form the extraction granularity.
    // example :
    //  vector.extract_strided_slice %s {
    //        offsets = [0, 0], sizes = [2, 2], strides = [1, 1]} : 
    //        vector<4x8x8xf32> to vector<2x2x8xf32>
    // here, extraction granularity is 8.
```

1. Please use consistent style in your comments (capitalisation).
2. Aavoid "random" line wrapping (e.g. in `sizes = [2, 2]`) -  that makes it hard to read.
3. This comment defines "granularity", but the code that follows computes `extractSliceLen`. Please make it more consistent with the implementation.

charithaintc replied:
renamed to `extractGranularitySize`. I think the work granularity captures the meaning well. so I like to keep it.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+    //  2],
+    //     strides = [1, 1]} : vector<4x8x8xf32> to vector<2x2x8xf32>
+    // here, extraction granularity is 8.
+    int64_t extractSliceLen = 1;
```
This represents "flattened slice len", right? Why not `flattenedSliceLen`? Naming is hard 🤷🏻

charithaintc replied:
addressed above. basically the extraction can happen only at first k dimensions of an n-D vectors. the remaining dimensions are extracted in full. so the last n-k dimensions form the extraction granularity. this is the meaning I wanted to capture in the naming.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+    int64_t extractSliceLen = 1;
+    auto n = extractOp.getSourceVectorType().getRank();
+    int64_t k = (int64_t)offsets.size();
+    if (n > k) {
```
Is this `if` is needed here? `k` is guaranteed to be ` <= n`, right?

charithaintc replied:
addressed above. n > k is true for the following example. 
``` 
%0 = vector.extract_strided_slice %arg0 { offsets = [1, 2], strides = [1, 1], sizes = [1, 4] }
    : vector<2x8x2xf32> to vector<1x4x2xf32>
```
In this case, only first 2 dims are extracted using the offsets, sizes etc. last dim is extracted in full. 
https://mlir.llvm.org/docs/Dialects/Vector/#vectorextract_strided_slice-vectorextractstridedsliceop

banach-space replied:
Right, so there are 2 cases:
* `k < n`
* `k = n`

right? I think that you can get rid of `if` by doing something like:
```cpp
int64_t outRank = (int64_t)offsets.size();
int64_t k = outRank;
while (k < n) {
  extractGranularitySize *=
            extractOp.getSourceVectorType().getShape()[outRank + (n-k)];
  ++k;
}
```

This way there's only one case. Also, I'd avoid variable names like `k` and `n` for anything other than loop iterators. Perhaps `numKDims` and `numMDims`? Or `inputRank`/`outputRank`? Or `kD`/`nD`? The last suggestion would be consistent with comments and already a nice improvement :)

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +106,236 @@ struct LinearizeVectorizable final
+    }
+    // Get total number of extracted slices.
+    int64_t nExtractedSlices = 1;
+    for (auto size : sizes) {
+      nExtractedSlices *= size.cast<IntegerAttr>().getInt();
+    }
```
Could you try using `computeSuffixProduct` instead?

charithaintc replied:
tried this. but this require constructing an `ArrayRef<int64_t>`. So I decided to use `llvm::for_each`.

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

charithaintc committed:
fix

charithaintc committed:
add comments

charithaintc committed:
fix

charithaintc committed:
fix

charithaintc committed:
fix comments

charithaintc commented:
Hi, @banach-space, @dcaballe, @Hardcode84. I think I have addressed all the comments. Can you please review the modifications and/or approve.

Thanks!

charithaintc committed:
fix comments

Hardcode84 commented:
LGTM, but wait for other reviewers

charithaintc committed:
fix

dcaballe commented on mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h:
```diff
@@ -389,6 +389,12 @@ void populateVectorLinearizeTypeConversionsAndLegality(
 
+/// Populates patterns for linearizing ND (N >= 2) vector operations to 1D
+/// vector shuffle operations.
+void populateVectorLinearizeToShuffleRewritePatterns(
```
Extract ops are "shuffle-like" ops so perhaps we can call this `populateVectorLinearizeShuffleLikeOpsPatterns` or similar

charithaintc replied:
renamed to `populateVectorLinearizeShuffleLikeOpsPatterns`

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
+    //      vector<4x8x8xf32> to vector<2x2x8xf32>
+    // Here, extraction granularity is 8.
+    int64_t extractGranularitySize = 1;
+    auto n = extractOp.getSourceVectorType().getRank();
```
nit: please, spell out the type when it's not literally redundant in the statement.

charithaintc replied:
fixed!

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
+    auto n = extractOp.getSourceVectorType().getRank();
+    int64_t k = (int64_t)offsets.size();
+    if (n > k) {
+      for (unsigned i = 0; i < n - k; i++) {
```
nit: use pre-increment per coding guidelines

charithaintc replied:
fixed.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
+    }
+    // Get total number of extracted slices.
+    int64_t nExtractedSlices = 1;
+    llvm::for_each(sizes, [&](Attribute size) {
```
`for (Attribute size : sizes) {...}` ?

charithaintc replied:
@banach-space recommended using `computeSuffixProduct`. But that does not take AttrArray as argument. So I am not sure to use it for this case. Changed back to a for loop for now.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
+          shuffleOp, "Can't flatten since targetBitWidth <= OpSize");
+
+    auto vec1 = adaptor.getV1();
+    auto vec2 = adaptor.getV2();
```
nit: spell out `auto`

charithaintc replied:
fixed.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorLinearize.cpp:
```diff
@@ -103,6 +108,252 @@ struct LinearizeVectorizable final
+    // outermost dim.
+    if (rank > 1) {
+      auto shape = shuffleOp.getV1().getType().getShape();
+      for (unsigned i = 1; i < shape.size(); i++) {
```
pre-increment

charithaintc replied:
fixed.

dcaballe reviewed (approved):
LGTM (module some nits that you can address before landing)

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

charithaintc committed:
address comments

banach-space reviewed (approved):
LGTM, left a few final suggestions, thanks!

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

charithaintc committed:
fix

charithaintc committed:
Merge branch 'main' into vector_linearize_mods

charithaintc commented:
Hi @Hardcode84, @banach-space, @dcaballe,

Thanks for the insightful reviews. This is my first PR in llvm-project and I learned a lot 😃 👍 

Can you please help me with the merge process.

Thanks!
